### PR TITLE
Service Discovery Service interface change

### PR DESCRIPTION
**Pull request by cw9**
@robskillington @xichen2020 @martin-mao

### CODE REVIEW

File: services/services.go [hunk 1]
```diff
@@ -53,8 +49,29 @@ type Service interface {
 	SetSharding(s ServiceSharding) Service
 }
 
-// NewService creates a new Service
-func NewService() Service { return new(service) }
+// PlacementSnapshot describes how shards are placed on instances
+type PlacementSnapshot interface {
```
Comment (robskillington) [inline services/services.go]: `ServicePlacement`?

File: services/services.go [hunk 2]
```diff
@@ -53,8 +49,29 @@ type Service interface {
...
+	SetReplication(r ServiceReplication) PlacementSnapshot
+
+	// Sharding returns the placement sharding description or nil if none
+	Sharding() ServiceSharding
+
+	// SetSharding sets the placement sharding description or nil if none
+	SetSharding(s ServiceSharding) PlacementSnapshot
+
+	// Validate checks if the placement is valid
+	Validate() error
```
Comment (robskillington) [inline services/services.go]: Do we need `Validate()` anymore if this is just a plain old object or POGO one might say?

File: services/services.go [hunk 3]
```diff
@@ -77,25 +91,27 @@ type ServiceSharding interface {
...
-	Endpoint() string                         // Endpoint address for contacting the instance
-	SetEndpoint(e string) ServiceInstance     // sets the endpoint address for the instance
-	Shards() shard.Shards                     // Shards owned by the instance
-	SetShards(s shard.Shards) ServiceInstance // sets the Shards assigned to the instance
+	PlacementInstance() string
+	SetPlacementInstance(i PlacementInstance) ServiceInstance
+	Service() string                         // the service implemented by the instance
+	SetService(s string) ServiceInstance     // sets the service implemented by the instance
+	Endpoints() []string                     // Endpoint addresses for contacting the instance
+	SetEndpoints(e []string) ServiceInstance // sets the endpoint addresses for the instance
```
Comment (robskillington) [inline services/services.go]: Let's keep a single endpoint per service instance as discussed.

File: services/services.go [hunk 4]
```diff
@@ -109,111 +125,52 @@ type Advertisement interface {
...
-	SetZones(zones []string) QueryOptions    // sets the list of zones to consult
 	IncludeUnhealthy() bool                  // if true, will return unhealthy instances
 	SetIncludeUnhealthy(h bool) QueryOptions // sets whether to include unhealthy instances
 }
 
-// NewQueryOptions creates new QueryOptions
-func NewQueryOptions() QueryOptions { return new(queryOptions) }
-
 // Services provides access to the service topology
 type Services interface {
```
Comment (robskillington) [inline services/services.go]: Let's add a method as discussed that returns a `PlacementService` for a given `query ServiceQuery` then each parameter on `PlacementService` already has the context of a service.

File: services/services.go [hunk 5]
```diff
@@ -109,111 +125,52 @@ type Advertisement interface {
...
+// all write or update operations will persist the generated snapshot before returning success
+type PlacementService interface {
+	BuildInitialPlacement(service ServiceQuery, hosts []PlacementInstance, shardLen int, rf int) (PlacementSnapshot, error)
+	AddReplica(service ServiceQuery) (PlacementSnapshot, error)
+	AddHost(service ServiceQuery, candidateHosts []PlacementInstance) (PlacementSnapshot, error)
+	RemoveHost(service ServiceQuery, host PlacementInstance) (PlacementSnapshot, error)
+	ReplaceHost(service ServiceQuery, leavingHost PlacementInstance, candidateHosts []PlacementInstance) (PlacementSnapshot, error)
+
+	// Snapshot gets the persisted snapshot for service
+	Snapshot(service ServiceQuery) (PlacementSnapshot, error)
```
Comment (robskillington) [inline services/services.go]: As discussed, would be good to have this `PlacementService` operate under the context of a given service already.

File: shard/shard.go [hunk 6]
```diff
@@ -20,10 +20,24 @@
 
 package shard
 
+const (
+	// INITIALIZING represents a shard newly assigned to an instance
+	INITIALIZING State = iota
```
Comment (robskillington) [inline shard/shard.go]: Let's stick with our current const naming scheme, this should just be `Initializing`.

File: shard/shard.go [hunk 7]
```diff
@@ -20,10 +20,24 @@
...
+
+	// AVAILABLE represends a shard bootstraped and ready to serve
+	AVAILABLE
```
Comment (robskillington) [inline shard/shard.go]: Let's stick with our current const naming scheme, this should just be `Available`.

File: services/services.go (hunk 2 above)
Comment (cw9) [inline services/services.go]: I'd still have Validate() here, cause we want to make sure every placement is validate before we save it, it's either here or a "validate(p ServicePlacement) error" somewhere in the persistent layer.

File: shard/shard.go (hunk 6 above)
Comment (cw9) [inline shard/shard.go]: will fix

File: services/services.go (hunk 1 above)
Comment (cw9) [inline services/services.go]: sure

**Pull request closed without merging**